Makefile

```diff
@@ -1,137 +1,205 @@
+DOCKER_RUN_73_IMAGE=devTools/Dockerfile-php73-xdebug.json
 
-.PHONY: all
-#Run all checks, default when running 'make'
-all: analyze test
+FLOCK=./devTools/flock
 
-#Non phony targets for phars etc.
-vendor: composer.json composer.lock
-	composer install
 
-build/cache:
-	mkdir -p build/cache
+#
+# Commands (phony targets)
+#---------------------------------------------------------------------------
 
-./php-cs-fixer-v2.phar:
-	wget $(PHP-CS-FIXER_URL)
-	chmod a+x ./php-cs-fixer-v2.phar
+.PHONY: compile
+compile:	## Bundles Infection into a PHAR
+compile: $(INFECTION)
 
-./phpstan.phar:
-	wget $(PHPSTAN_URL)
-	chmod a+x ./phpstan.phar
+.PHONY: cs-fix
+cs-fix:	  	## Runs PHP-CS-Fixer
+cs-fix: $(PHP_CS_FIXER)
+	# TODO: disable verbose mode?
+	$(PHP_CS_FIXER) fix -v --cache-file=$(PHP_CS_FIXER_CACHE)
 
-#All tests, (infection itself, phpunit, e2e) for different php version/ environments (xdebug or phpdbg)
-.PHONY: test test-unit test-infection-phpdbg test-e2e-phpdbg test-infection-xdebug test-e2e-xdebug
-test: test-unit test-infection-phpdbg test-e2e-phpdbg test-infection-xdebug test-e2e-xdebug
-	# All tests finished without errors
+.PHONY: cs-check
+cs-check:	## Runs PHP-CS-Fixer in dry mode
+cs-check: $(PHP_CS_FIXER)
+	# TODO: disable verbose mode?
+	$(PHP_CS_FIXER) fix -v --cache-file=$(PHP_CS_FIXER_CACHE) --dry-run --stop-on-violation
 
-.PHONY: test-unit test-unit-72 test-unit-73
-#php unit tests
-test-unit: test-unit-72 test-unit-73
+.PHONY: phpstan
+phpstan:  	## Runs PHPStan
+phpstan: vendor $(PHPSTAN)
+	$(PHPSTAN) analyse src --level=max --configuration ./devTools/phpstan-src.neon --no-interaction --no-progress
+	$(PHPSTAN) analyse tests --level=4 --configuration ./devTools/phpstan-tests.neon --no-interaction --no-progress
 
-test-unit-72: build-xdebug-72
-	$(DOCKER_RUN_72) $(PHPUNIT)
+.PHONY: analyze
+analyze:	## Runs CS fixers, static analyzers and various other checks
+analyze: cs-check analyze-ci
 
-test-unit-73: build-xdebug-73
-	$(DOCKER_RUN_73) $(PHPUNIT)
+.PHONY: analyze-ci
+analyze-ci:	## Runs static analyzers and various other checks
+analyze-ci: phpstan validate
 
-.PHONY: test-infection-phpdbg test-infection-phpdbg-72 test-infection-phpdbg-73
-#infection with phpdbg
-test-infection-phpdbg: test-infection-phpdbg-72 test-infection-phpdbg-73
+.PHONY: auto-review
+auto-review:	## TODO: find a suitable description
+auto-review: vendor
+	$(PHPUNIT) --group=auto-review
 
-test-infection-phpdbg-72: build-xdebug-72
-	$(DOCKER_RUN_72) phpdbg -qrr bin/infection --threads=4
+.PHONY: validate
+validate:	## Checks that the composer.json file is valid
+validate:
+	composer validate --strict
 
-test-infection-phpdbg-73: build-xdebug-73
-	$(DOCKER_RUN_73) phpdbg -qrr bin/infection --threads=4
+.PHONY: test
+test:		## Runs all the tests
+# TODO: add a test to ensure we are not missing targets here
+test: test-unit test-e2e-phpdbg test-e2e-xdebug test-infection-phpdbg test-infection-xdebug
+
+.PHONY: test-unit
+test-unit:	## Runs the unit tests
```

Comment:
A suggestion here. As of now you will see (in the make help command) the:

```
test-unit: ...
test-unit-72: ...
test-unit-73: ...
```

same for all others. What do you think instead of "hidding" (you just need to remove the command line) the `7*` ones? As a result you would only show:

```
test-unit: ...
test-e2e: ...
test-infection: ...
```

The commands don't disappear, but they are less highlighted because more of a detail that you will figure out when running the above commands if necessary. I.e. what you care about is `test-unit`, if `test-unit-72` in peculiar fails and you want to run only this one, you will discover that command when the failure occur and will be free to run this more specific command if desired

Reply:
I think there's a way to make dynamic rules (like, e.g. with `$(PHP_CS_FIXER)` rule below), so *theoretically* all these rules like `test-unit-72` and `test-unit-73` can be compacted into one. But, well, this will make things less explicit, and, not necessarily easier to understand, so I didn't think into this approach too far.

Reply:
Yeah, I'm fine with having it like that and try not too be too smart about since it will make the Makefile less accessible for not much.

The point above though was more about the displayed command when running `make` rather than the actual targets.